### ISSUE TITLE
reduce_sum == "sum" missing scaling factor

### ISSUE THREAD

**Issue by romitjain** [allenai/open-instruct#995]
Hi,

I am looking into the implementation of `reduce_sum = loss`. According to me, the loss should be reduced using sum reduction and then scaled based on the total number of tokens in the batch.

This is also the gist of the description [here](https://github.com/allenai/open-instruct/blob/main/open_instruct/finetune.py#L793-L799).

The linked issue in the comments and the following _correct_ implementation in transformers handles it [correctly](https://github.com/huggingface/transformers/blob/fbeaf96f9e2291c21277ac658a33ea8752728bf3/src/transformers/loss/loss_utils.py#L41).

However, the same is not getting reflected in open instruct. The loss is [summed](https://github.com/allenai/open-instruct/blob/main/open_instruct/finetune.py#L814) but never divided by the total tokens in the batch.

Even while logging, the metrics report a much lower loss after scaling appropriately [here](https://github.com/allenai/open-instruct/blob/main/open_instruct/finetune.py#L910-L918).

Is this expected or am I missing something?

**Comment by finbarrtimbers** [comment]
Ok, so there's a bit of a weird history behind this feature. 

Originally, we didn't scale by the number of tokens (as you've pointed out). We then added support to our mean loss implementation to track the number of tokens. Since our mean loss implementation now does track the number of tokens correctly, I have a [PR](https://github.com/allenai/open-instruct/pull/1024) out now to remove the (incorrect) sum loss implementation.

**Comment by romitjain** [comment]
@finbarrtimbers 
I am curious. Were the tulu3 models trained using this sum loss implementation?

**Comment by finbarrtimbers** [comment]
@hamishivi would know better than me!

**Comment by hamishivi** [comment]
Tulu 3 models were yes! We describe the setup in section 4.3.2.

**Comment by romitjain** [comment]
Thanks, @finbarrtimbers and @hamishivi, for the clarification.

Apologies, but I wanted to clarify one more thing: 
Usually when we scale to multinode or multi GPU training, we multiply the loss by `num_processes` too, due to implicit scaling of optimizer states by `num_processes` as mentioned:
1. [here](https://huggingface.co/docs/accelerate/en/usage_guides/gradient_accumulation)
> We multiply by num_processes because the DDP calculates the average gradient across all devices whereas dividing by num_items_in_batch already takes into account all devices
2. [here](https://github.com/huggingface/transformers/blob/main/src/transformers/trainer.py#L4145), when gradient accumulation is switched on in `SFTTrainer`

Both of them are only requried in case of gradient accumulation. Open instruct does not do the same. What are your thoughts on this?

**Comment by hamishivi** [comment]
@romitjain great question! You are right that we have this implicit scaling per device with ddp. With the older implementation, this basically just meant that we sum per-device and then average across devices. I guess we could have scaled up by the number of processes, but I didn't think to do this.

**Comment by romitjain** [comment]
Thank you @hamishivi for the clarification!
